### ISSUE TITLE
Control regressions since rebase

### ISSUE THREAD

**Issue by bslenul** [libretro/dolphin#163]
**edit:** Recap of the regressions since the rebase:

* ~~L and R triggers not being analog anymore~~ ✅ fixed with #166 (it also adds the possibility to bind "half-presses" now 👍 )
* ~~L1/R1 and L2/R2 swapped by default~~ ✅ fixed with #166
* ~~Device type set to "RetroPad" by default in GC mode with empty binds ("---")~~ ✅ fixed with #166 (it also sets "Wiimote" as default device type in Wii mode)
* ~~Left mouse not working as an "A" press anymore~~ ✅ fixed with #226
* ~~"Wiimote IR" options not existing anymore~~ ✅ fixed with #229
* ~~"Real Wiimote" not working anymore~~ ✅ fixed with #242
* ~~No matter if ports 2-4 are set to "GameCube Controller", only port 1 will work, because only `SIDevice0` is correctly set to "6" (= GC controller) in Dolphin.ini, `SIDevice1/2/3` are stuck at "0" (= None)~~ ✅ fixed with #246
* ~~Not possible to use a GC controller in Wii mode in some games anymore (Super Smash Bros. Brawl and Dragon Ball Z - Budokai Tenkaichi 3 for example)~~ ✅ fixed with #246

---
Original post:

Hey, I was testing latest commit (f4f4947 atm) on both Windows and Linux and noticed that triggers don't work properly anymore, now it only "full press" L/R directly, which is really undesired in some games (Mario Sunshine, Metroid Prime, etc.), it worked fine with 1fbd599 for example (which was the latest successful build available on Windows before the buildbot hack), the core was able to make the distinction between slight and full presses.

Maybe unrelated but I noticed that now L and R are mapped by default to L1 and R1 instead of L2 and R2 (and Z is now mapped to R2 instead of R1), and in `Quick Menu > Controls` it doesn't say "GameCube Controller" anymore but "RetroPad".

**Comment by Mr-Jake** [comment]
This was the PR that fixed analog triggers in the dolphin core. Not sure if it made it into the recent rebase.   

https://github.com/libretro/dolphin/pull/59

**Comment by bslenul** [comment]
Ah yup, just tried on Windows and I can confirm it fixes the main issue 👍 But yeah, looks like the core is getting a lot of work recently, maybe it's just not ready yet and I probably shouldn't have opened this issue so prematurely... I'll stay patient and stick with 1fbd599 until it's ready! Closing this for now.

**Comment by Mr-Jake** [comment]
Maybe leave it open so @aliaspider is aware of it and can re-commit flyinghead's PR https://github.com/libretro/dolphin/pull/59.

Thanks for confirming that the PR fixes the issue.

**Comment by bslenul** [comment]
Sure, I'll reopen :D 

Also yeah, the Controls menu is a bit messy atm, like I said previously it shows "RetroPad" instead of "GameCube Controller", and everything is set to `---` until you switch device type:

![image](https://user-images.githubusercontent.com/33353403/91966854-b5973000-ed12-11ea-8e6c-48a4de89696b.png)

And after switching there's a "Select" button and L and R are mapped to L1 and R1 by default, and Z is mapped to R2.

But again, maybe the devs are already working on it and this is a bit premature to report :p

**Comment by i30817** [comment]
Does this have anything to do with this:

https://github.com/libretro/dolphin/issues/69

In that bug certain actions were not working in Eternal Darkness if controlled by the keyboard; because they were mapped by the *game* to analog buttons (but they only required digital full button press). One of them was aiming or shooting the weapon iirc; the other was switching screens in the options/inventory menu (though it's been a while).

Apparently, the mapping had to change so the gamecube 'analog' button matched a 'analog' button on the retropad that then was mapped to a fully digital keyboard button and finally then the game thought the analog button was fully depressed. There was some wizardry to distinguish between a analog or digital button and making the 'digital' take precedence if the controller button is digital (such as in a keyboard).

This is the commit: https://github.com/libretro/dolphin/pull/75/commits/d7c6a6187e43cd3f76bb900f3909796fe13afb08

This seems like the more recent commit affecting this area than the one you linked.

**Comment by bslenul** [comment]
The issue with the triggers in current commit is that they're assumed to be fully pressed directly every time you press them. So it's either "not pressed" or "fully pressed", no in-between. Which is exactly like using a keyboard for L and R, it's either fully pressed or not pressed at all. So in Mario Sunshine for example, there's no way to run and use the water gadget at the same time as a "half press" is required for that.

For the commit idk, #59 worked for me but I could try #75 as well.

For keyboard, IMO, the ideal would be to add "L-Analog" and "R-Analog" in the Controls menu, that's what standalone does: 
![image](https://user-images.githubusercontent.com/33353403/92106396-9e287780-ede4-11ea-9799-15b135bc4525.png)

So basically: if you have both "L" and "L-Analog" bound to the same button it will uses analog properly, but if you bound a different button/key for each, then both actions will be separated.

No idea how hard it would be to implement tho...

**Comment by i30817** [comment]
Honestly i prefer the current behavior.... err, the last behaviour that worked after 75.

If the controller has analog capability, it's treated as analog, if it does not, it's treated as digital. It limits the need to have different mappings if you're using a controller or a keyboard, just adapts the input to the capabilities of the device.

warning: this is what my idea of what is supposed to happen is. I only noticed the bug that 75 fixes *because* i don't have a controller and use a keyboard, so it's perfectible possible that commit broke analog and i didn't notice, though i'm confident that jdgleaver would.

**Comment by bslenul** [comment]
Heh I don't know, it can be really punitive in some games to have "all or nothing", a good example is Metroid Prime: when you hold L button it's supposed to lock your aim (with no enemies on front of you I mean) which allows you to strafe for example, and while messing with latest commit I noticed:

* Half press L = aim locked.
* Full press with analog trigger (so half press -> full press) = aim locked.
* Full press directly, WITHOUT half press first = no lock!

So on keyboard or without analog triggers you can't strafe and/or lock your aim in a specific position. Locking on enemies, scanning, etc. still work perfectly fine tho.

And like I said in Mario Sunshine where you can't run while using your water gadget if you're unable to half press. And there are probably more games using that half/full press distinction.

> It limits the need to have different mappings if you're using a controller or a keyboard

That wouldn't really be an issue if it adds new "L/R (half press)" actions for example to Controls menu, defaults would still be set to "L" and "R" which would be analog if your controller supports it, or full press if digital. So default behavior wouldn't change at all, but you'd be able to bind half presses if you want/need.

**Comment by i30817** [comment]
That's a definite point, if at least if the 'normal' retropad triggers keep their adaptability of analog when on a analog device and digital on a digital device, and leave the 'half pressed' empty and ready to be setup for particular games.

But maybe we could try to ask for the 'fullpress' of a analog-button-on-digital-device be first 'half press' for some frames (or even one considering how most input loops work) to help games recognize what's going on? Might make these half-pressed 'extras' unnecessary for the games that glitch on digital full-press right away, if they don't explicitly depend on being 'half-pressed', but it's probably a good idea anyway for the digital device 'analog fake'.

**Comment by bslenul** [comment]
@RobertTheSable I just tested your PR, seems to fix most of the issues I mentioned in this thread ❤️ 

RetroPad is now shown properly again as "GameCube Controller", L/R bound by default to L2/R2 (with analog support), etc. and L-Analog/R-Analog for half presses (bound to L3/R3 by default) is an awesome addition for non-analog users 👍 

However! Binding L/R to a digital key/button doesn't do anything, so to keep Mario Sunshine as an example it's impossible to strafe (full L press) or to lock Mario in place to use the water gadget (full R press). L/R-Analog work fine if bound to digital key/button tho!

**Comment by RobertTheSable** [comment]
> RetroPad is now shown properly again as "GameCube Controller", L/R bound by default to L2/R2 (with analog support), etc. and L-Analog/R-Analog for half presses (bound to L3/R3 by default) is an awesome addition for non-analog users +1
> 

Realized this just after I opened the PR. I have a fix prepared to allow the fullpresses to be rebound, but in my testing it doesn't resolve the issue of the trigger input being hardcoded(i.e. binding R2 to A still results in an R-Trigger pressing Dolphin), so I'd like to see if I can resolve that.

**Comment by RobertTheSable** [comment]
Being unable to unmap the triggers appears to not be a Dolphin-specific issue(I noticed the same issue in Flycast), so I pushed code that should at least let you map hard presses to another button.

**Comment by bslenul** [comment]
Oh yeah, same issue with Flycast indeed oO Never remapped the triggers before so never noticed...

Tested your updated PR, works like a charm! Thank you! 👍

**Comment by bslenul** [comment]
Found a pretty bad one with Super Mario Galaxy 1 and 2: if using a real Wiimote with "continuous scanning" core option it just hangs RetroArch :( 

**edit:** I guess it has something to do with the Bluetooth connection since it doesn't hang if the Wiimote is physically OFF.

**Comment by bslenul** [comment]
Hey, just noticed while talking with someone on Reddit that newer versions of the core lack the "Wiimote IR" core options (screen from 1fbd599):

![image](https://user-images.githubusercontent.com/33353403/104757908-ed50c280-575d-11eb-8419-a583da57b512.png)

The "Wiimote IR Mode" option especially is REALLY useful since it let you point at the screen without the need of a mouse...

**edit:** Also with the older version of the core you could use the left mouse click as an A press, which was really useful as well when using a mouse as pointer.

**Comment by bslenul** [comment]
Updated OP with a quick recap! Tell me if I missed something, I'll add it to the list.

**Comment by bslenul** [comment]
Added #217 to the list.

**Comment by Mr-Jake** [comment]
Looking at the code from before the rebase:

"Wiimote IR" core options
\dolphin-master\Source\Core\DolphinLibretro\Options.cpp

```

Option<int> irMode("dolphin_ir_mode", "Wiimote IR Mode", 1,
                     {"Right Stick controls pointer (relative)", "Right Stick controls pointer (absolute)", "Mouse controls pointer"});
Option<int> irCenter("dolphin_ir_center", "Wiimote IR Center",
  { {"50", 50}, {"60", 60}, {"70", 70}, {"80", 80}, {"90", 90}, {"100", 100}, {"0", 0}, {"10", 10}, {"20", 20}, {"30", 30}, {"40", 40} });
Option<int> irWidth("dolphin_ir_width", "Wiimote IR Width",
  { {"50", 50}, {"60", 60}, {"70", 70}, {"80", 80}, {"90", 90}, {"100", 100}, {"0", 0}, {"10", 10}, {"20", 20}, {"30", 30}, {"40", 40} });
Option<int> irHeight("dolphin_ir_height", "Wiimote IR Height",
  { {"50", 50}, {"60", 60}, {"70", 70}, {"80", 80}, {"90", 90}, {"100", 100}, {"0", 0}, {"10", 10}, {"20", 20}, {"30", 30}, {"40", 40} });

```

I assume the rest of the code needed would be in
\dolphin-master\Source\Core\Core\HW\


Could these control regressions be resolved by using portions of the old code?

**Comment by ds22x** [comment]
"No matter if ports 2-4 are set to "GameCube Controller", only port 1 will work, because only SIDevice0 is correctly set to "6" (= GC controller) in Dolphin.ini, SIDevice1/2/3 are stuck at "0" (= None)"
I assume is why it's currently impossible to use a GameCube controller scheme for Wii games that support it (as it needs to be set to port 5 and up)?

**Comment by bslenul** [comment]
I don't think that's the issue, having a GC controller in port 5 works fine in Bomberman Blast or Mario Kart Wii for example:

![image](https://user-images.githubusercontent.com/33353403/132747996-f0ea669f-c957-4410-95de-47ca4f49b491.png)

![image](https://user-images.githubusercontent.com/33353403/132746464-cbe4cf19-289d-42fa-a7c7-18f0e69d57de.png)

However it doesn't work if you have only the GC controller plugged (device type to "None" for ports 1-4).

And it doesn't seem to work with every game, I can't get it to work in Super Smash Bros. Brawl or Dragon Ball Z - Budokai Tenkaichi 3, so I'm a bit confused 🤔

**Comment by ds22x** [comment]
If it helps, the only way I got a GC controller to work with Brawl was to have a Dolphin specific controller override in the game's .ini file (circumventing the LibRetro controller api).

**Comment by bslenul** [comment]
Added to the list.

Seems to be because by default the GC controller is set to `Libretro/4/Joypad`, and some games don't like that, when "hardcoding" `Libretro/0/Joypad` it works fine in Brawl.

Not sure how that could be handled, maybe a "8 Players (Wii Mode)" core option, when OFF (default) it would re-enable "GameCube Controller" as possible device type for ports 1-4 and disable ports 5-8, and when ON ports 1-4 would only show Wii device types and GC controller for ports 5-8 (just like it is right now).

**Comment by bslenul** [comment]
Fix submitted few days ago for the last 2 remaining issues, if anyone wants to try (link to the .dll and .so files in the PR if you don't know how or can't compile the core yourself).

**Comment by ds22x** [comment]
Seems to work just fine.
